controllers/raycluster_controller.go

```diff
@@ -0,0 +1,320 @@
+	routeapply "github.com/openshift/client-go/route/applyconfigurations/route/v1"
+	routev1client "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
+)
+
+// RayClusterReconciler reconciles a RayCluster object
+type RayClusterReconciler struct {
+	client.Client
+	kubeClient  *kubernetes.Clientset
+	routeClient *routev1client.RouteV1Client
+	Scheme      *runtime.Scheme
+	CookieSalt  string
+}
+
+const (
+	requeueTime             = 10
+	controllerName          = "codeflare-raycluster-controller"
+	oauthAnnotation         = "codeflare.dev/oauth=true"
+	CodeflareOAuthFinalizer = "codeflare.dev/oauth-finalizer"
+	OAuthServicePort        = 443
+	OAuthServicePortName    = "oauth-proxy"
+	strTrue                 = "true"
+	strFalse                = "false"
+	logRequeueing           = "requeueing"
+)
+
+var (
+	deletePolicy  = metav1.DeletePropagationForeground
+	deleteOptions = client.DeleteOptions{PropagationPolicy: &deletePolicy}
+)
+
+//+kubebuilder:rbac:groups=ray.io,resources=rayclusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ray.io,resources=rayclusters/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ray.io,resources=rayclusters/finalizers,verbs=update
+//+kubebuilder:rbac:groups=route.openshift.io,resources=routes,verbs=patch;delete;get
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get;create;patch;delete;get
+//+kubebuilder:rbac:groups=core,resources=services,verbs=patch;delete;get
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=patch;delete;get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the RayCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.3/pkg/reconcile
+
+func (r *RayClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	var cluster rayv1.RayCluster
+
+	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Error getting RayCluster resource")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&cluster, CodeflareOAuthFinalizer) {
+			logger.Info("Add a finalizer", "finalizer", CodeflareOAuthFinalizer)
+			controllerutil.AddFinalizer(&cluster, CodeflareOAuthFinalizer)
+			if err := r.Update(ctx, &cluster); err != nil {
+				logger.Error(err, "Failed to update RayCluster with finalizer", logRequeueing, strTrue)
+				return ctrl.Result{Requeue: true, RequeueAfter: requeueTime}, err
```

Comment:
It seems based on documentation that the `Requeue` bool is not necessary to be set as it is implied when using `RequeueAfter`. It could be removed on all instances.

https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.3/pkg/reconcile#Result